### PR TITLE
[DTable] Tables.jl - dtable as a source

### PR DESCRIPTION
**Pull request by krynju**
- [x] row access (enough for CSV)
- [x] column access without iterators (enough to work with DataFrames)
- [x] column access iterator
- [x] schema (necessary for Arrow)
- [x] tests
- [x] cleanup 
- [x] gdtable things

### CODE REVIEW

**Comment by krynju** [comment]
@quinnj Hey, this is a PR with things required to make the DTable a Tables.jl source.

Please note the set of constructors might be a little messy at the moment (might have to readjust the existing ones as part of this PR).
A grouped dtable struct will be introduced in a different PR, so the Tables.jl code for it will eventually find it's way in here as well,

Everything seems to work well with the major tables.jl consumers already, so it's looking good to me

Please have a look at this PR at your convenience and let me know what you think!

File: src/table/dtable.jl [hunk 1]
```diff
@@ -19,12 +19,32 @@ the underlying partitions was applied to it (currently only `filter`).
...
+    DTable(chunks::VTYPE, tabletype, schema) = new(chunks, tabletype, schema)
 end
 
 DTable(chunks::Vector{Dagger.EagerThunk}, args...) = DTable(VTYPE(chunks), args...)
 DTable(chunks::Vector{Dagger.Chunk}, args...) = DTable(VTYPE(chunks), args...)
 
+"""
+    DTable(table; tabletype=nothing) -> DTable
+
+Constructs a `DTable` using a `Tables.jl` compatible `table` input.
```
Comment (jpsamaroo) [inline src/table/dtable.jl:33]:
```suggestion
Constructs a `DTable` using a `Tables.jl`-compatible input `table`.
```

File: src/table/dtable.jl [hunk 2]
```diff
@@ -19,12 +19,32 @@ the underlying partitions was applied to it (currently only `filter`).
...
+Calls `Tables.partitions` on the `table` and assumes the provided partitioning.
```
Comment (jpsamaroo) [inline src/table/dtable.jl:34]:
```suggestion
Calls `Tables.partitions` on `table` and assumes the provided partitioning.
```

File: src/table/dtable.jl [hunk 3]
```diff
@@ -19,12 +19,32 @@ the underlying partitions was applied to it (currently only `filter`).
...
+"""
+function DTable(table; tabletype=nothing)
+    chunks = Vector{Dagger.Chunk}()
+    type = nothing
+    sink = Tables.materializer(tabletype !== nothing ? tabletype() : partition)
+    for partition in Tables.partitions(table)
+        tpart = sink(partition)
+        push!(chunks, Dagger.tochunk(tpart))
+        type === nothing && (type = typeof(tpart).name.wrapper;)
```
Comment (jpsamaroo) [inline src/table/dtable.jl:43]:
```suggestion
        if type === nothing
            type = typeof(tpart).name.wrapper
        end
```

Makes it clear that you're not trying to construct a `NamedTuple` :smile:

File: src/table/dtable.jl [hunk 4]
```diff
@@ -36,26 +56,16 @@ argument to partition the table (based on row count).
...
-        tpart = sink(p)
-        push!(chunks, Dagger.tochunk(tpart))
-        if type === nothing
-            type = typeof(tpart).name.wrapper
+    sink = Tables.materializer(tabletype !== nothing ? tabletype() : table)
+    for outer_partition in Tables.partitions(table)
+        for inner_partition in Tables.partitions(TableOperations.makepartitions(outer_partition, chunksize))
+            tpart = sink(inner_partition)
+            push!(chunks, Dagger.tochunk(tpart))
+            type === nothing && (type = typeof(tpart).name.wrapper;)
```
Comment (jpsamaroo) [inline src/table/dtable.jl:66]:
```suggestion
            if type === nothing
                type = typeof(tpart).name.wrapper
            end
```

File: src/table/tables.jl [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+
+
+#######################################
+# DTableRowIterator functions
+
+Tables.schema(table::DTableRowIterator) = Tables.schema(table.d)
+length(table::DTableRowIterator) = length(table.d)
+
+function Base.iterate(iter::DTableRowIterator; c_idx=0)
```
Comment (jpsamaroo) [inline src/table/tables.jl:73]: Can you document what `c_idx` does here?

File: src/table/tables.jl [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+# DTableColumnIterator functions
+
+Tables.schema(table::DTableColumnIterator) = Tables.schema(table.d)
+Tables.columnnames(table::DTableColumnIterator) = Tables.columnnames(table.d)
+Tables.getcolumn(table::DTableColumnIterator, col::Symbol) = Tables.getcolumn(table.d, col)
+Tables.getcolumn(table::DTableColumnIterator, idx::Int) = Tables.getcolumn(table.d, idx)
+length(table::DTableColumnIterator) = length(Tables.columnnames(table))
+
+
+function Base.iterate(table::DTableColumnIterator; idx=1)
```
Comment (jpsamaroo) [inline src/table/tables.jl:105]: Docs for `idx` kwarg

File: src/table/tables.jl [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+
+
+#######################################
+# DTablePartitionIterator functions
+
+Tables.partitions(table::DTable) = DTablePartitionIterator(table)
+length(table::DTablePartitionIterator) = length(table.d.chunks)
+
+function Base.iterate(table::DTablePartitionIterator; idx=1)
```
Comment (jpsamaroo) [inline src/table/tables.jl:122]: Docs for `idx` kwarg

**Review by jpsamaroo — APPROVED** [review]
Awesome stuff!

File: src/table/tables.jl (hunk 5 above)
Comment (krynju) [reply src/table/tables.jl:73]: I'm just going to redo these iterator kwargs. Was speedrunning the iterators and this let me write fewer functions, but it isn't elegant now that I look at it

**Comment by jpsamaroo** [comment]
Thanks again!

**Pull request merged**